packages/dds/shared-object-base/src/sharedObjectBase.ts

```diff
@@ -0,0 +1,448 @@
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).bindToContext}
+     */
+    public bindToContext(): void {
+        if (this._isBoundToContext) {
+            return;
+        }
+
+        this._isBoundToContext = true;
+
+        // Allow derived classes to perform custom processing prior to registering this object
+        this.registerCore();
+
+        this.runtime.bindChannel(this);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).connect}
+     */
+    public connect(services: IChannelServices) {
+        this.services = services;
+        this.attachDeltaHandler();
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).isAttached}
+     */
+    public isAttached(): boolean {
+        return this.services !== undefined && this.runtime.attachState !== AttachState.Detached;
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).captureState}
+     */
+    public captureState(): IStateHost {
+        const state = this.captureStateCore();
+        return { summarize: async (fullTree: boolean = false) => {
+            return state.summarize(this._serializer, fullTree);
+        }};
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).getGCData}
+     */
+    public async getGCData(fullGC: boolean = false): Promise<IGarbageCollectionData> {
+        // Set _isGCing to true. This flag is used to ensure that we only use SummarySerializer (created in
+        // getGCDataCore) to serialize handles in this object's data.
+        // Assertions for re-entrancy are only to ensure _isGCing value is correct. If we want to support re-entrancy
+        // in the future, it can be changed to a count of GC calls in progress.
+        assert(!this._isGCing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isGCing = true;
+
+        let gcData: IGarbageCollectionData;
+        try {
+            gcData = await this.getGCDataCore();
+            assert(this._isGCing, 0x079 /* "Possible re-entrancy! Summary should have been in progress." */);
+        } finally {
+            this._isGCing = false;
+        }
+
+        return gcData;
+    }
+
+    /**
+     * Returns the GC data for this shared object. It contains a list of GC nodes that contains references to
+     * other GC nodes.
+     * Derived classes must override this to provide custom list of references to other GC nodes.
+     */
+    protected async getGCDataCore(): Promise<IGarbageCollectionData> {
+        // We run the full summarize logic to get the list of outbound routes from this object. This is a little
+        // expensive but it's okay for now. It will be updated to not use full summarize and make it more efficient.
+        // See: https://github.com/microsoft/FluidFramework/issues/4547
+        const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
+        const state = this.captureStateCore();
+        await state.summarize(serializer, false);
```

Comment:
By the time SharedObject captured state, it's too late to give it a serializer. I think the right model is the following:
- captureStateCore (or captureState, not sure there is need for both here) would internally create a serializer and capture snapshot with it, keeping serializer
- The object returned has summarize() method that does not take serializer, and simply returns a summary.
- There is also another method, called getGCData(), that would use internally stored serializer to spit it out.

That way serializer is left to be implementation detail, and API is relatively high level:
- get a revision (aka state)
- ask revision to generate summary.
- ask revision to generate GC info

Reply:
As @agarwal-navin [comments](https://github.com/microsoft/FluidFramework/pull/8422#discussion_r770752289), running GC can change what should be in the summary. Given that the potential change is only to say whether something is referenced or not, it seems possible to just fix that up in the IChannelRevision. That feels like a strange thing to do though.